File: archive/seema-node/src/routes/remediation.ts
import { Router, Request, Response } from 'express';
import prisma from '../lib/prisma.js';
import { authenticate } from '../middleware/auth.js';
import { getTenantFilter } from '../middleware/tenant.js';
import { logAudit } from '../middleware/auditLogger.js';
import { z } from 'zod';

const router = Router();

// Serialize Prisma's camelCase to the snake_case shape the frontend expects.
// `steps` is stored as a JSON string — parse before returning. `assigned_to`
// and `due_date` come straight from the columns of the same name.
function serializeRemediationPlan(p: any) {
  let steps: unknown[] = [];
  try {
    steps = p.steps ? JSON.parse(p.steps) : [];
  } catch { /* leave as empty array */ }
  // Derived counters for the UI.
  const stepsArr = Array.isArray(steps) ? steps as any[] : [];
  const totalSteps = stepsArr.length;
  const completedSteps = stepsArr.filter(
    (s: any) => s?.status === 'completed' || s?.completed === true,
  ).length;
  return {
    id: p.id,
    title: p.title,
    source: p.source,
    source_type: p.source,           // legacy alias
    priority: p.priority,
    status: p.status,
    assigned_to: p.assignedTo,
    due_date: p.dueDate,
    steps,
    total_steps: totalSteps,
    completed_steps: completedSteps,
    created_at: p.createdAt,
    updated_at: p.updatedAt,
  };
}

// GET /compliance/remediation-plans — List remediation plans for firm
router.get('/compliance/remediation-plans', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const plans = await prisma.remediationPlan.findMany({
      where: { firmId },
      orderBy: { createdAt: 'desc' },
    });
    res.json(plans.map(serializeRemediationPlan));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch remediation plans' });
  }
});

// POST /compliance/remediation-plans — Create remediation plan
//
// Required: title, priority. `steps` is optional — the frontend often creates
// a plan from an alert with no pre-set steps; users then add steps later.
// `alert_id` and `assigned_to` are accepted but currently stored only as
// metadata on the plan (no foreign-key column for alert_id yet).
const createPlanSchema = z.object({
  title: z.string(),
  priority: z.string(),
  status: z.string().optional(),
  // Description / source_id are accepted but not persisted (no columns yet).
  description: z.string().nullable().optional(),
  source_id: z.string().nullable().optional(),
  due_date: z.string().nullable().optional(),
  // snake_case (frontend)
  alert_id: z.string().nullable().optional(),
  assigned_to: z.string().nullable().optional(),
  source_type: z.string().nullable().optional(),
  // camelCase (legacy)
  source: z.string().optional(),
  alertId: z.string().nullable().optional(),
  assignedTo: z.string().nullable().optional(),
  dueDate: z.string().nullable().optional(),
  steps: z.array(z.object({
    title: z.string(),
    description: z.string().optional(),
    completed: z.boolean().optional(),
  })).optional(),
});

router.post('/compliance/remediation-plans', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const parsed = createPlanSchema.safeParse(req.body);
    if (!parsed.success) {
      res.status(400).json({ error: true, message: 'Validation failed', details: parsed.error.errors });
      return;
    }

    const { title, priority } = parsed.data;
    const source = parsed.data.source ?? parsed.data.source_type ?? null;
    const assignedTo = parsed.data.assigned_to ?? parsed.data.assignedTo ?? null;
    const dueDateStr = parsed.data.due_date ?? parsed.data.dueDate ?? null;
    const status = parsed.data.status ?? 'pending';
    const steps = parsed.data.steps ?? [];

    const plan = await prisma.remediationPlan.create({
      data: {
        firmId,
        title,
        source,
        priority,
        assignedTo,
        dueDate: dueDateStr ? new Date(dueDateStr) : null,
        status,
        steps: JSON.stringify(steps),
      },
    });

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'remediation_plan_created',
      entityType: 'remediation_plan',
      entityId: plan.id,
      metadata: {
        title,
        priority,
        alert_id: parsed.data.alert_id ?? parsed.data.alertId ?? null,
        assigned_to: parsed.data.assigned_to ?? parsed.data.assignedTo ?? null,
      },
    });

    res.status(201).json(serializeRemediationPlan(plan));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to create remediation plan' });
  }
});

// POST /compliance/remediation-steps/:stepId/complete — Complete a step in a plan
router.post('/compliance/remediation-steps/:stepId/complete', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const { stepId } = req.params as Record<string, string>;
    // Frontend sends snake_case (plan_id); older callers sent camelCase
    // (planId). Accept both.
    const planId = (req.body.plan_id ?? req.body.planId) as string | undefined;

    if (!planId) {
      res.status(400).json({ error: true, message: 'plan_id (or planId) is required' });
      return;
    }

    const plan = await prisma.remediationPlan.findFirst({
      where: { id: planId, firmId },
    });
    if (!plan) {
      res.status(404).json({ error: true, message: 'Remediation plan not found' });
      return;
    }

    let steps: Array<{ title: string; description?: string; completed?: boolean; id?: string }> = [];
    try {
      steps = JSON.parse(plan.steps || '[]');
    } catch {
      res.status(500).json({ error: true, message: 'Failed to parse plan steps' });
      return;
    }

    const stepIndex = parseInt(stepId, 10);
    if (isNaN(stepIndex) || stepIndex < 0 || stepIndex >= steps.length) {
      res.status(404).json({ error: true, message: 'Step not found' });
      return;
    }

    steps[stepIndex].completed = true;

    const updated = await prisma.remediationPlan.update({
      where: { id: planId },
      data: { steps: JSON.stringify(steps) },
    });

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'remediation_step_completed',
      entityType: 'remediation_plan',
      entityId: planId,
      metadata: { stepId, stepTitle: steps[stepIndex].title },
    });

    res.json(serializeRemediationPlan(updated));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to complete remediation step' });
  }
});

// POST /compliance/remediate — Resolve a remediation plan
//
// Frontend sends snake_case (plan_id); older callers sent camelCase
// (planId). Accept both.
const remediateSchema = z.object({
  // snake_case (frontend)
  plan_id: z.string().optional(),
  // camelCase (legacy)
  planId: z.string().optional(),
}).refine(
  (d) => Boolean(d.plan_id || d.planId),
  { message: 'plan_id (or planId) is required' },
);

router.post('/compliance/remediate', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const parsed = remediateSchema.safeParse(req.body);
    if (!parsed.success) {
      res.status(400).json({ error: true, message: 'Validation failed', details: parsed.error.errors });
      return;
    }

    const planId = (parsed.data.plan_id ?? parsed.data.planId) as string;

    const plan = await prisma.remediationPlan.findFirst({
      where: { id: planId, firmId },
    });
    if (!plan) {
      res.status(404).json({ error: true, message: 'Remediation plan not found' });
      return;
    }

    const updated = await prisma.remediationPlan.update({
      where: { id: planId },
      data: { status: 'resolved' },
    });

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'remediation_plan_resolved',
      entityType: 'remediation_plan',
      entityId: planId,
    });

    res.json(serializeRemediationPlan(updated));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to resolve remediation plan' });
  }
});

export default router;

File: backend/routers/remediation.py
"""Remediation plan tracking router."""
import uuid
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from middleware.tenant_rls import tenant_db_from_jwt, bypass_db
from middleware.auth import get_current_user, CurrentUser
from services.audit_logger import log_audit
from models.remediation import RemediationPlan

router = APIRouter()

class RemediationPlanCreate(BaseModel):
    title: str
    source: str = None
    priority: str = "medium"
    assigned_to: str = None
    due_date: datetime = None
    steps: str = None

@router.get("/compliance/remediation-plans")
async def list_remediation_plans(
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """List remediation plans."""
    stmt = select(RemediationPlan).where(
        RemediationPlan.firm_id == current_user.firm_id
    )
    result = await db.execute(stmt)
    plans = result.scalars().all()
    return [
        {
            "id": p.id,
            "firm_id": p.firm_id,
            "title": p.title,
            "source": p.source,
            "priority": p.priority,
            "status": p.status,
            "assigned_to": p.assigned_to,
            "due_date": p.due_date,
            "steps": p.steps,
            "created_at": p.created_at,
            "updated_at": p.updated_at,
        }
        for p in plans
    ]

@router.post("/compliance/remediation-plans")
async def create_remediation_plan(
    payload: RemediationPlanCreate,
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Create remediation plan."""
    plan = RemediationPlan(
        id=str(uuid.uuid4()),
        firm_id=current_user.firm_id,
        title=payload.title,
        source=payload.source,
        priority=payload.priority,
        assigned_to=payload.assigned_to,
        due_date=payload.due_date,
        steps=payload.steps,
        status="pending",
    )
    db.add(plan)
    await db.flush()
    await log_audit(
        db=db,
        firm_id=current_user.firm_id,
        action="created",
        entity_type="remediation_plan",
        entity_id=plan.id,
        user_id=current_user.user_id,
        details=f"Created remediation plan: {plan.title}",
    )
    return {
        "id": plan.id,
        "firm_id": plan.firm_id,
        "title": plan.title,
        "source": plan.source,
        "priority": plan.priority,
        "status": plan.status,
        "assigned_to": plan.assigned_to,
        "due_date": plan.due_date,
        "steps": plan.steps,
        "created_at": plan.created_at,
        "updated_at": plan.updated_at,
    }

@router.post("/compliance/remediation-steps/{step_id}/complete")
async def complete_remediation_step(
    step_id: str,
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Mark remediation step complete."""
    stmt = select(RemediationPlan).where(
        RemediationPlan.firm_id == current_user.firm_id
    )
    result = await db.execute(stmt)
    plan = result.scalar_one_or_none()
    if not plan:
        raise HTTPException(status_code=404, detail="Not found")
    await log_audit(
        db=db,
        firm_id=current_user.firm_id,
        action="completed_step",
        entity_type="remediation_plan",
        entity_id=plan.id,
        user_id=current_user.user_id,
        details=f"Completed step {step_id}",
    )
    return {"status": "success", "step_id": step_id}

class MarkResolvedRequest(BaseModel):
    plan_id: str

@router.post("/compliance/remediate")
async def mark_plan_resolved(
    req: MarkResolvedRequest,
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Mark remediation plan resolved."""
    stmt = select(RemediationPlan).where(
        RemediationPlan.id == req.plan_id,
        RemediationPlan.firm_id == current_user.firm_id,
    )
    result = await db.execute(stmt)
    plan = result.scalar_one_or_none()
    if not plan:
        raise HTTPException(status_code=404, detail="Not found")
    plan.status = "completed"
    await db.flush()
    await log_audit(
        db=db,
        firm_id=current_user.firm_id,
        action="resolved",
        entity_type="remediation_plan",
        entity_id=plan.id,
        user_id=current_user.user_id,
        details=f"Marked plan resolved: {plan.title}",
    )
    return {
        "id": plan.id,
        "status": plan.status,
        "completed_at": datetime.utcnow(),
    }
